### PR TITLE
No lib contract

### PR DESCRIPTION
**Pull request by LayneHaber**
Contract improvements + redeployments:

- Multisend interpreter pattern
- Changes to data structures for gas optimizations
- Changes to emitted events
- Deployments to rinkeby/goerli

### CODE REVIEW

File: modules/contracts/README.md [hunk 1]
```diff
@@ -1,19 +1,59 @@
...
+1. The `prepare` call above emits an event with the same calldata. Router hears this event (which includes its address) and calls the `prepare` function with the same calldata on the receiving chain (with decremented fees and timeout).
+1. User hears `prepare` event on receiver side. User validates the event data and signs the data, then broadcasts it to the network.
+1. Any router can then submit `fulfill` on the receiver side using the user's commitment + event data. If they do so, they get a small fee (in addition to gas).
+1. Original router, upon seeing the `fulfill` event on receiver side, submits `fulfill` to sender side.
+1. Note that in both sender and receiver side cases, `fulfill` must be called before the timeout expires. This acts as a failsafe against funds getting locked indefinitely if the counterparty is malicious. However, this also means expiry must be far enough away (w/ enough gap between both sides) to _make sure_ the tx will go through.
```
Comment (jakekidd) [inline modules/contracts/README.md:15]: All of these are step 1 ?

File: modules/contracts/README.md [hunk 2]
```diff
@@ -1,19 +1,59 @@
...
+
+Once the proper environment variables are added to your environment, you can begin the contract deployments by running the following from the `modules/contracts` directory:
+
+```sh
+npx hardhat deploy --network "<NETWORK_NAME_FOR_CHAIN_ID>" # e.g. npx hardhat deploy --network "goerli"
+```
+
+You should use the `NETWORK_NAME` that corresponds to the correct network within the `hardhat.config.ts` file.
+
+**NOTE:** Once you have deployed the contracts, you will then need to update (if necessary) and redeploy the subgraphs. See [here](https://github.com/connext/nxtp/tree/main/modules/subgraph) for details.
```
Comment (jakekidd) [inline modules/contracts/README.md:59]: Thank you, this is an awesome README

File: modules/contracts/README.md (hunk 1 above)
Comment (LayneHaber) [reply modules/contracts/README.md:15]: No that's next to fix lol

File: modules/contracts/README.md (hunk 2 above)
Comment (LayneHaber) [reply modules/contracts/README.md:65]: (ignoring the 18 steps 1)

**Pull request merged**